### PR TITLE
Bug 1142540 - move contacts.js to promises.

### PR DESCRIPTION
**Pull request by julienw**

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a5e6828296a84d8646af5981bd892975a7bb0244

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e6868bc068b9066a8018fffc2abea11d5d3266a7

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=15c318b3eaa7af0880d8a61397b5c6e874d6ce34

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3d58cc0c0297ea1a17bb6e8750547c48e36b3263

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a638e7cc646e4bd32aa728eca69b904ac62bf847

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7df717b3c7cf435da3d9f64d11518228a3c02688

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=12b797f6af63496032a5538272e29b8d8d121c3f

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=12b797f6af63496032a5538272e29b8d8d121c3f

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=12b797f6af63496032a5538272e29b8d8d121c3f

File: apps/sms/js/thread_ui.js [hunk 1]
```diff
@@ -688,38 +688,40 @@ var ThreadUI = {
...
+      data && this.recipients.add(data);
+
+      Compose.fromMessage(activity);
+
+      if (number) {
+        Compose.focus();
+      } else {
+        this.recipients.focus();
+      }
+    });
```
Comment (julienw) [inline apps/sms/js/thread_ui.js]: here I rewrote this part with a IMO cleaner and more readable promise-based code, tell me if this is ok

File: apps/sms/js/utils.js [hunk 2]
```diff
@@ -504,9 +508,6 @@
 
       // Only exit when no record and no phone number case.
       if (!record && !number) {
-        if (typeof callback === 'function') {
-          callback(null);
-        }
```
Comment (julienw) [inline apps/sms/js/utils.js]: was never used, I think this was a left over from old logic

File: apps/sms/test/unit/thread_ui_test.js [hunk 3]
```diff
@@ -1267,22 +1271,28 @@ suite('thread_ui.js >', function() {
...
-        test('input value is invalid ', function() {
+        test('input value is invalid ', function(done) {
           // An actual accepted recipient from contacts
           fixture.number = 'foo';
           fixture.isQuestionable = true;
 
           ThreadUI.recipients.add(fixture);
-          assert.isFalse(fixture.isInvalid);
+          /* The previous add triggers an asynchronous behavior when validating
+           * this new recipient. Let's wait for this. */
```
Comment (julienw) [inline apps/sms/test/unit/thread_ui_test.js]: I think we need to find a better way to test this; ideally I'd like that MockRecipients is dumber and especially does not emit events. But this is really painful to change now and I think it's not the goal of this patch... so here is a workaround.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=9bf50c23b8d59f139734d98485e7add59c8cdc47

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2790a6e448ceacc0b775846548c9c11640b94005

File: apps/sms/js/contacts.js [hunk 4]
```diff
@@ -98,29 +98,26 @@
...
+         *    early avoids the additional overhead of making a
+         *    request that we know will fail.
+         *
+         * 2. Missing navigator.mozContacts API or empty but
+         *    present filter.filterValue make the app useless,
+         *    so prevent unpredictable state by making the
+         *    result a no-op.
+         *
+         */
+        return Promise.resolve(
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: Do we really need this ```null``` return value? I'm not sure if we have code somewhere that has drastically different behavior for empty array and null result.

IMO it just makes code that consumes API more complex - we always should check ```contacts && contacts.length``` or ```contacts || []``` like in methods below. Usually list-based API methods always return an array. In our case if we have really unrecoverable error we should just reject promise.

What do you think?

File: apps/sms/js/contacts.js [hunk 5]
```diff
@@ -206,31 +200,27 @@
...
-      };
-
-      request.onerror = function onerror() {
-        // When an error occurs, regardless of completed count,
-        // clear the onsuccess handler from this request and immediately
-        // invoke the callback with a `null` argument.
-        this.onsuccess = this.onerror = null;
-        callback(null);
-      };
+        return results;
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: I see you improved this code a bit (removed slice), so just on the side note - we can improve this callback further with for example:
```
if (terms.length > 1) {
  var fields = ['tel', 'givenName', 'familyName'];
  if (Settings.supportEmailRecipient) {
    fields.push('email');
  }
  var criteria = { fields: fields, terms: lower };
  return contacts.filter((contact) => isMatch(contact, criteria, filterFns.contains));
}

return contacts;
```

No need to do it in the patch - just if you want :)

File: apps/sms/js/contacts.js [hunk 6]
```diff
@@ -206,31 +200,27 @@
...
+      }).catch(() => {
+        return null;
+      });
     },
-    findByString: function contacts_findBy(filterValue, callback) {
+
+    findByString: function contacts_findBy(filterValue) {
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: nit: ```function contacts_findByString```

File: apps/sms/js/contacts.js [hunk 7]
```diff
@@ -206,31 +200,27 @@
...
       var unknownCallback = function(contacts) {
         contacts = contacts || [];
-        this.findByUnknown(filterValue, function(unknown) {
-          unknown = unknown || [];
-          callback(contacts.concat(unknown));
-        });
+        return this.findByUnknown(filterValue)
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: Maybe I'm biased, but if we use ```Promise.all``` to run this _find_ in "parallel" the code will look more readable:
```
return Promise.all([
  this.findContactByString(filterValue),
  this.findByUnknown(filterValue)
]).then(function(contacts) {
  var knownContacts = contacts[0];
  var unknownContacts = contacts[1];

  return unknownContacts.length > 0 ?
    knownContacts.concat(unknownContacts) :
    knownContacts;
});
```
What do you think?

File: apps/sms/js/contacts.js [hunk 8]
```diff
@@ -206,31 +200,27 @@
...
+            return contacts;
+          });
       }.bind(this);
-      this.findContactByString(filterValue, unknownCallback);
+
+      return this.findContactByString(filterValue).then(unknownCallback);
     },
 
-    findContactByString: function contacts_findBy(filterValue, callback) {
+    findContactByString: function contacts_findBy(filterValue) {
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: nit: ```function contacts_findContactByString```

File: apps/sms/js/contacts.js [hunk 9]
```diff
@@ -239,9 +229,10 @@
         filterBy: props,
         filterOp: 'contains',
         filterValue: filterValue
-      }, callback);
+      });
     },
-    findByUnknown: function findByUnknown(filterValue, callback) {
+
+    findByUnknown: function findByUnknown(filterValue) {
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: nit: ```function contacts_findByUnknown``` to be consistent with the rest of methods

File: apps/sms/js/contacts.js [hunk 10]
```diff
@@ -258,16 +249,16 @@
           list.push(obj);
         }
       }
-      callback(list);
+      return Promise.resolve(list);
     },
 
-    findExact: function contacts_findBy(filterValue, callback) {
+    findExact: function contacts_findBy(filterValue) {
```
Comment (azasypkin) [inline apps/sms/js/contacts.js]: nit: ```function contacts_findExact```

File: apps/sms/js/desktop-only/navigator_moz_contacts.js [hunk 11]
```diff
@@ -217,53 +217,14 @@
...
-                  onerror.call(this);
-                  onerror = null;
-                }.bind(this));
-              }
-            }
-          }
-        }
+      // Implement a mock that gives results that appear to
+      // match the real behaviour of filtered contacts results
+      var result =  ContactsDB.filter(function(contact) {
```
Comment (azasypkin) [inline apps/sms/js/desktop-only/navigator_moz_contacts.js]: nit: trailing space

File: apps/sms/js/information.js [hunk 12]
```diff
@@ -438,18 +438,18 @@ Information.prototype = {
...
-      Contacts.findByAddress(number, (results) => {
+      Contacts.findByAddress(number).then((contacts) => {
         // If the current rendering ID doesn't match the latest ID, skip current
         // one and only render for ID which matches latest request ID.
         if (currentRenderingId !== this.renderingId) {
           return;
         }
 
-        var isContact = results !== null && !!results.length;
+        var isContact = contacts !== null && !!contacts.length;
```
Comment (azasypkin) [inline apps/sms/js/information.js]: If we always return an array - here will be just onliner nice check - ```if (contacts.length)``` :)

File: apps/sms/js/thread_ui.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: I definitely OK with this! But honestly I don't see any value in ```Utils.getContactDisplayInfo```, it's used only here, internally it basically does nothing except that it calls public ```Utils.basicContact``` that's why I got rid of it in my patch (including useless tests, because in reality ```Utils.basicContact``` should be tested instead). + we use this approach in ThreadUI.handleDraft :) Do you think it's still useful?

File: apps/sms/js/thread_ui.js [hunk 13]
```diff
@@ -734,10 +736,10 @@ var ThreadUI = {
       // Recipients will exist for draft messages in threads
       // Otherwise find them from draft recipient numbers
       draft.recipients.forEach(function(number) {
-        Contacts.findByAddress(number, function(records) {
-          if (records.length) {
+        Contacts.findByAddress(number).then(function(contacts) {
+          if (contacts.length) {
```
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: Can we potentially get ```null reference``` exception here with the current patch(as before patch as well)?

File: apps/sms/js/thread_ui.js [hunk 14]
```diff
@@ -2513,17 +2512,18 @@ var ThreadUI = {
 
     if (event.target.isPlaceholder) {
       typed = event.target.textContent.trim();
-      this.searchContact(typed, this.listContacts.bind(this));
+      this.searchContact(typed)
```
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: nit: indentation looks weird, or it's only me :)

I'd rather do like this:
```
this.searchContact(typed).then(
  (contacts) => this.listContacts(typed, contacts)
);
```

File: apps/sms/js/thread_ui.js [hunk 15]
```diff
@@ -2513,17 +2512,18 @@ var ThreadUI = {
...
+        .then((contacts) => this.listContacts(typed, contacts));
     }
 
     this.emit('recipientschange');
   },
 
-  exactContact: function thui_searchContact(fValue, handler) {
-    Contacts.findExact(fValue, handler.bind(null, fValue));
+  exactContact: function thui_searchContact(fValue) {
```
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: nit: ```function thui_exactContact```

File: apps/sms/js/thread_ui.js [hunk 16]
```diff
@@ -2531,10 +2531,10 @@ var ThreadUI = {
       // eg. type "a", then delete it.
       // Always remove the the existing results.
       this.toggleRecipientSuggestions();
-      return;
+      return Promise.resolve(null);
```
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: Looks like here you actually changed program flow a bit for the case when everything was deleted from recipients input, earlier it was like this: ```toFieldInput -> searchContact -> toggleRecipientSuggestions```, now it's like this: ```toFieldInput -> searchContact -> toggleRecipientSuggestions -> listContacts```. Probably it doesn't have any consequences yet, but why don't you want use ```Promise.reject()``` here - sounds like a reasonable use case for reject here?

File: apps/sms/js/thread_ui.js [hunk 17]
```diff
@@ -2735,9 +2735,9 @@ var ThreadUI = {
       tel = number || '';
     }
 
-    Contacts.findByAddress(number, function(results) {
-      var isContact = results && results.length;
-      var contact = results[0];
+    Contacts.findByAddress(number).then(function(contacts) {
```
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: nit: why not arrow function here as you like? :)

File: apps/sms/js/utils.js (hunk 2 above)
Comment (azasypkin) [inline apps/sms/js/utils.js]: Agree, I did the same in my patch :)

File: apps/sms/test/unit/thread_ui_test.js [hunk 18]
```diff
@@ -5925,7 +5942,9 @@ suite('thread_ui.js >', function() {
         MessageManager.on.withArgs('message-received').yield({
           message: testMessage
         });
-      }).then(done, done);
+
+        return Contacts.findByPhoneNumber.lastCall.returnValue;
```
Comment (azasypkin) [inline apps/sms/test/unit/thread_ui_test.js]: Could you please add comment that explains why we use ```return Contacts.findByPhoneNumber.lastCall.returnValue;``` here?

File: apps/sms/test/unit/thread_ui_test.js [hunk 19]
```diff
@@ -5225,7 +5239,9 @@ suite('thread_ui.js >', function() {
       Navigation.isCurrentPanel.withArgs('composer').returns(true);
 
       // Please, auto-answering stub, don't interfere with my test
-      this.sinon.stub(Contacts, 'findExact');
+      this.sinon.stub(Contacts, 'findExact').returns({
+        then: () => {}
```
Comment (azasypkin) [inline apps/sms/test/unit/thread_ui_test.js]: nit: maybe ```Promise.resolve(...)```?

File: apps/sms/test/unit/thread_ui_test.js (hunk 3 above)
Comment (azasypkin) [inline apps/sms/test/unit/thread_ui_test.js]: Sure, I'm fine such workaround in this patch

File: apps/sms/test/unit/mock_contacts.js [hunk 20]
```diff
@@ -3,53 +3,21 @@
...
-
-  addUnknown: () => {},
-
-  mTeardown: function mc_mTeardown() {
-    this.mAsync = false;
+    return Promise.resolve(result);
   }
-};
+
+  exports.MockContacts = {
```
Comment (azasypkin) [inline apps/sms/test/unit/mock_contacts.js]: yay, much cleaner now!

File: apps/sms/test/unit/information_test.js [hunk 21]
```diff
@@ -112,50 +113,61 @@ suite('Information view', function() {
...
-          number: participants[0]
-        });
-        assert.isTrue(!!reportView.contactList.firstElementChild);
+
+        Contacts.findByAddress.lastCall.returnValue.then(() => {
+          sinon.assert.notCalled(ContactRenderer.prototype.render);
+          sinon.assert.calledWith(Template.prototype.interpolate, {
+            number: participants[0]
+          });
+          assert.isTrue(!!reportView.contactList.firstElementChild);
```
Comment (azasypkin) [inline apps/sms/test/unit/information_test.js]: nit: ```assert.isNotNull```? It was here previously, so I'm ok if you leave it as it was before.

File: apps/sms/test/unit/contacts_test.js [hunk 22]
```diff
@@ -467,56 +393,49 @@ suite('Contacts', function(done) {
...
+        // navigator.mozContacts.find was called?
+        assert.equal(mHistory.length, 1);
+      }).then(done, done);
     });
 
     test('string search yields a contact without familyName', function(done) {
-      Contacts.findContactByString('julien 123', function(contacts) {
-        done();
-      });
+      Contacts.findContactByString('julien 123').then(() => done(), done);
```
Comment (azasypkin) [inline apps/sms/test/unit/contacts_test.js]: Weird test, tbh :) What does it test?

File: apps/sms/test/unit/contacts_test.js [hunk 23]
```diff
@@ -39,156 +39,108 @@ suite('Contacts', function(done) {
...
+          result = null;
+        }
+
+        // All other cases
+
+        if (result === null) {
+          return Promise.reject({
+            name: 'Mock missing filter params'
+          });
+        } else {
```
Comment (azasypkin) [inline apps/sms/test/unit/contacts_test.js]: nit: JSLint would say you have redundant ```else``` block here :) JSHint is more liberal I think

File: apps/sms/test/unit/contacts_test.js [hunk 24]
```diff
@@ -39,156 +39,108 @@ suite('Contacts', function(done) {
...
+
+        // Supports the error case
+        if (filter == null ||
+            (!filter.filterOp || !filter.filterValue)) {
+          result = null;
+        }
+
+        if (filter.filterValue === targetLocalNumber ||
+            filter.filterValue === targetLocalEmail) {
+          result = [{
```
Comment (azasypkin) [inline apps/sms/test/unit/contacts_test.js]: nit: ideally we should have ```else if``` here, but since it's test code I'm fine with the current solution as well

File: apps/sms/js/thread_ui.js (hunk 1 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: I think I agree with you here; IIRC that was something the former owner insisted on so I didn't want to push too much against it.

File: apps/sms/js/contacts.js (hunk 5 above)
Comment (julienw) [inline apps/sms/js/contacts.js]: good idea, did it :)

File: apps/sms/js/contacts.js (hunk 7 above)
Comment (julienw) [inline apps/sms/js/contacts.js]: I did a somewhat different version of yours, you'll tell me what you think :)

File: apps/sms/js/thread_ui.js (hunk 1 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: but I'd rather do it in a separate patch

File: apps/sms/js/contacts.js (hunk 4 above)
Comment (julienw) [inline apps/sms/js/contacts.js]: I agree but i'd rather do this in a separate patch because we can clean up a lot of places and I think it's not really part of the migration to promises.

File: apps/sms/js/information.js (hunk 12 above)
Comment (julienw) [inline apps/sms/js/information.js]: agreed but let's do that in a separate bug :)

File: apps/sms/js/thread_ui.js (hunk 13 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: theorically but the cases where we get "null" are really specific; I don't think this can really happen here. That said I agree we should just get rid of this possibility.

File: apps/sms/js/thread_ui.js (hunk 14 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: this works in this case, but not "generally" if we need to add more .then handlers. But I can do this here :)

File: apps/sms/js/thread_ui.js (hunk 16 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: yes, I think it's a goood idea !

File: apps/sms/js/thread_ui.js (hunk 17 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: because it was not previously done :) i'd rather not do it then

File: apps/sms/test/unit/thread_ui_test.js (hunk 18 above)
Comment (julienw) [inline apps/sms/test/unit/thread_ui_test.js]: mmm i'm not so sure it's needed after all

Comment (julienw) [inline apps/sms/test/unit/thread_ui_test.js]: but I'm quite sure I saw it broken without it, so I'll leave it and add some comment; maybe it would be an intermittent failure without this.

File: apps/sms/test/unit/thread_ui_test.js (hunk 19 above)
Comment (julienw) [inline apps/sms/test/unit/thread_ui_test.js]: no, I especially wanted the "then" handler to do nothing; but maybe I need to return another thenable that does nothing?

File: apps/sms/js/contacts.js (hunk 4 above)
Comment (azasypkin) [inline apps/sms/js/contacts.js]: Sure, just don't forget to file a follow-up for it :)

File: apps/sms/js/contacts.js (hunk 7 above)
Comment (azasypkin) [inline apps/sms/js/contacts.js]: okay :)

File: apps/sms/js/information.js (hunk 12 above)
Comment (azasypkin) [inline apps/sms/js/information.js]: Agreed in the comment above :)

File: apps/sms/js/thread_ui.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: Okay

File: apps/sms/js/thread_ui.js (hunk 14 above)
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: Yeah, it would be great if you add the case with several ```then``` to https://etherpad.mozilla.org/sms-code-style-guidelines as we have a lot of them now and will have more in the future :)

File: apps/sms/js/thread_ui.js (hunk 17 above)
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: Okay, both ways work for me. I just noticed that you changed some functions to arrow functions that is why I was wondering how that one was different :)

File: apps/sms/test/unit/contacts_test.js (hunk 22 above)
Comment (julienw) [inline apps/sms/test/unit/contacts_test.js]: actually, it was originally testing that the callback as actually called by in that case. This is not happening if I remove the code https://github.com/mozilla-b2g/gaia/blob/130079ebdb3821275c2e5d5152a70686e2dd133b/apps/sms/js/contacts.js#L60-L62

However with the patch the error is now caught by the "catch" in https://github.com/julienw/gaia/blob/2790a6e448ceacc0b775846548c9c11640b94005/apps/sms/js/contacts.js#L204-L206...

I changed the test to assert that the object is not null, but now I think I should instead remove the "catch" I added and keep a rejected promise. What do you think ?

File: apps/sms/test/unit/thread_ui_test.js (hunk 19 above)
Comment (azasypkin) [inline apps/sms/test/unit/thread_ui_test.js]: >  to do nothing

You mean you need kind-of-promise that never resolves? If so, maybe then ```new Promise(()=>{})```?

I don't really like that we change return signature for ```findExact``` that potentially can lead to a missed bug in the code

File: apps/sms/test/unit/contacts_test.js (hunk 24 above)
Comment (julienw) [inline apps/sms/test/unit/contacts_test.js]: ideally we should have switch/case blocks :)

File: apps/sms/js/thread_ui.js (hunk 16 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: So I did it but I'm still not sure it's a good idea; to me, rejection should really be errors (as in "exception"), but here it's a totally normal and expected case. So i'd rather keep a resolved Promise after all.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7ef282585579dfd8106f085ae99eeb9e93c3cc29

File: apps/sms/js/thread_ui.js (hunk 16 above)
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: Well, I partly agree, I have mixed feelings, IMO here is ```exception``` case since we don't pass correct arguments to ```searchContacts```, but we have another problem here ```toggleRecipientSuggestions``` call in ```searchContacts``` - doesn't look like it's correct place for such call. You asking to search for something, but it can change UI :)

File: apps/sms/test/unit/contacts_test.js (hunk 24 above)
Comment (azasypkin) [inline apps/sms/test/unit/contacts_test.js]: Yeah maybe, big ```switch\case```'s always scare me and indentation is too deep :)

File: apps/sms/test/unit/contacts_test.js (hunk 22 above)
Comment (azasypkin) [inline apps/sms/test/unit/contacts_test.js]: You now I like ```Promise.reject```, so I'd prefer that :) But I'm not sure everything will work as expected until we check it thoroughly :) Maybe leave it for the follow-up.

Comment (julienw) [inline apps/sms/test/unit/contacts_test.js]: well, in the previous code, a similar case would not invoke the callback; with a rejection I think the same happens. So it would be quite similar here.

"catch" without effect is just like a `catch(e){}`: useless :)

I'm changing this

File: apps/sms/test/unit/thread_ui_test.js (hunk 19 above)
Comment (julienw) [inline apps/sms/test/unit/thread_ui_test.js]: tell me if the new code looks better for you, otherwise I can do your promise-that-never-resolves.

File: apps/sms/js/thread_ui.js (hunk 16 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: I removed toggleRecipientSuggestions so that it's run in listContacts now (I moved it inside listContacts as well).

I think this is not an exception because this happens in normal use of the app; do you think we should just not call searchContact in case there is no filterValue, and remove the suggestion panel earlier in the code?

i like the current code (as with my patch), tell me what you think.

File: apps/sms/test/unit/contacts_test.js (hunk 24 above)
Comment (julienw) [inline apps/sms/test/unit/contacts_test.js]: actually, we should just use sinon from the test themselves rather than this :p

I'll file a follow up for this too.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2834c6acb807a416d4df4457ef46f7ad16dce14d

File: apps/sms/test/unit/contacts_test.js (hunk 24 above)
Comment (azasypkin) [inline apps/sms/test/unit/contacts_test.js]: Fair enough :)

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=7dd3dbddd324f5ef752887f253c6e50cf7ed07cf

File: apps/sms/test/unit/contacts_test.js (hunk 24 above)
Comment (julienw) [inline apps/sms/test/unit/contacts_test.js]: https://bugzilla.mozilla.org/show_bug.cgi?id=1153883

File: apps/sms/js/contacts.js (hunk 4 above)
Comment (julienw) [inline apps/sms/js/contacts.js]: https://bugzilla.mozilla.org/show_bug.cgi?id=1153888

File: apps/sms/js/thread_ui.js (hunk 14 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: done

File: apps/sms/js/thread_ui.js (hunk 1 above)
Comment (julienw) [inline apps/sms/js/thread_ui.js]: https://bugzilla.mozilla.org/show_bug.cgi?id=1153885

File: apps/sms/js/thread_ui.js (hunk 16 above)
Comment (azasypkin) [inline apps/sms/js/thread_ui.js]: > ; do you think we should just not call searchContact in case there is no filterValue, and remove the suggestion panel earlier in the code?

Yep, something like this sounds reasonable to me

File: apps/sms/js/contacts.js (hunk 7 above)
Comment (azasypkin) [inline apps/sms/js/contacts.js]: Yep, your version is better. ```([known, unknown])``` - didn't know that it's possible, with this ```Promise.all``` callbacks will be so sweet! And ```(known || [])``` is more correct since we still can return ```null``` :)

File: apps/sms/test/unit/information_test.js [hunk 25]
```diff
@@ -167,16 +179,21 @@ suite('Information view', function() {
...
-        });
-        assert.isTrue(
-          !!reportView.contactList.firstElementChild.firstElementChild);
+
+        Contacts.findByAddress.lastCall.returnValue.then(() => {
+          sinon.assert.notCalled(ContactRenderer.prototype.render);
+          sinon.assert.calledWith(Template.prototype.interpolate, {
+            number: participants[0].number
+          });
+          assert.isTrue(
```
Comment (azasypkin) [inline apps/sms/test/unit/information_test.js]: nit: ```assert.isNotNull```, but I don't block on this anyway :)

File: apps/sms/test/unit/information_test.js [hunk 26]
```diff
@@ -198,38 +222,48 @@ suite('Information view', function() {
...
-          sinon.assert.notCalled(Template.prototype.interpolate);
-          assert.equal(reportView.contactList.textContent, '');
-          Contacts.findByAddress.lastCall.yield([]);
-
-          // Only the new participant rendered on the contactList.
-          sinon.assert.calledWith(Template.prototype.interpolate, {
-            number: newParticipant[0].number
-          });
-          assert.equal(reportView.contactList.querySelectorAll('li').length, 1);
+          defer1.resolve({ contacts: [] });
```
Comment (azasypkin) [inline apps/sms/test/unit/information_test.js]: typo: ```{ contacts: [] }``` ---> ```[]```

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8a9d1f9fc89da932207c27105d97882c4c0b6de1

**Comment by autolander** [comment]
http://docs.taskcluster.net/tools/task-graph-inspector/#PVX9vDT3SuavQpI5CiVQ3g

The pull request failed to pass integration tests. It could not be landed, please try again.
